submodules/islandora_webform_ingest/includes/utilities.inc

```diff
@@ -496,40 +503,35 @@ function iwi_set_form_element_byhash(&$form, &$form_state, &$cloned_elements, $e
             case "tabpanel":
               $tab = $element->findElement($child_element_hash);
               if ($tab) {
-                if (!array_key_exists($child_element_hash, $cloned_elements)) {
-                  // If not already cloned clone it.
-                  $new_tab = clone $tab;
-
-                  $set_element_properties = function ($element) {
-                    $element->default_value = NULL;
-                  };
-                  // Set element properties on the descendants.
-                  $new_tab->eachDecendant($set_element_properties);
-
-                  $element->adopt($new_tab);
-                  $new_tab_drupal = $new_tab->toArray();
-
-                  drupal_array_set_nested_value($new_tab_drupal, array_merge($future_partial_path, array('#attributes')), $input_value['#attributes']);
-                  drupal_array_set_nested_value($new_tab_drupal, array_merge($future_partial_path, array('#default_value')), $input_value['#default_value']);
-
-                  $special_element_ref[] = $new_tab_drupal;
-
-                  // Appending the drupal form array back to where it belongs.
-                  // Lets store where do we put this one, so we can later just
-                  // fill out the values if we need to.
-                  $special_element_children = element_children($special_element_ref);
-                  $cloned_elements[$child_element_hash] = end($special_element_children);
+                $special_element_children = element_children($special_element_ref);
+                if (!isset($special_element_children[$offset])) {
+                  $lastkid = end($special_element_children);
+                  for ($i = $lastkid; $i < $offset; $i++) {
```

Comment:
Pat, but we have to be able to increment those, what happens if some presses "add" button. He gets a new one but with a numeric index, more over, the previous one gets replaced by a 0.
I will take that special case in account, not like the idea of getting involved in root-id:tabs, makes things really complicated to parse(pre parse) and post parse…we don't have that structure in $form.

Lest see that tomorrow.

Best


Diego Pino Navarro
Krayon Media
Pedro de Valdivia 575
Pucón - Chile
F:+56-45-2442469




On Oct 8, 2015, at 3:10 PM, Pat Dunlavey <notifications@github.com> wrote:

> In submodules/islandora_webform_ingest/includes/utilities.inc:
> 
> > -                  $new_tab_drupal = $new_tab->toArray();
> > -
> > -                  drupal_array_set_nested_value($new_tab_drupal, array_merge($future_partial_path, array('#attributes')), $input_value['#attributes']);
> > -                  drupal_array_set_nested_value($new_tab_drupal, array_merge($future_partial_path, array('#default_value')), $input_value['#default_value']);
> > -
> > -                  $special_element_ref[] = $new_tab_drupal;
> > -
> > -                  // Appending the drupal form array back to where it belongs.
> > -                  // Lets store where do we put this one, so we can later just
> > -                  // fill out the values if we need to.
> > -                  $special_element_children = element_children($special_element_ref);
> > -                  $cloned_elements[$child_element_hash] = end($special_element_children);
> > +                $special_element_children = element_children($special_element_ref);
> > +                if (!isset($special_element_children[$offset])) {
> > +                  $lastkid = end($special_element_children);
> > +                  for ($i = $lastkid; $i < $offset; $i++) {
> Take a look at the attached screenshots. One is from the "MODS Universal" form that Williams created; the other is from a copy of the "Basic Image MODS" form. It turns out that tabpanels do not necessarily have numeric ids. More importantly, the array index for tabpanels in the form schema is not equivalent to instances of tabpanels in an actual form. It would be more accurate to define paths something like this:
> root-id:tabs-id:tabpanel-id[0]:fieldset-id:etc. rather than root-id:tabs-id:tabpanel-id:fieldset-id. So the offset should be relative to a node in the path, it should not be the node.
> 
> Real world examples:
> name:0[0]:namePart
> name:0[1]:namePart
> nameTabs:namePanel[0]:nameSet:namePart
> nameTabs:namePanel[1]:nameSet:namePart
> 
> As for how to implement this and how our current implementation needs to change, thought needed!
> 
> 
> 
> —
> Reply to this email directly or view it on GitHub.
>